infra/pySourceCredChart/js/lib/pySourceCredChart.js

```diff
@@ -0,0 +1,287 @@
+var widgets = require('@jupyter-widgets/base');
+
+import * as d3 from 'd3';
+
+var sourceCredChartModel = widgets.DOMWidgetModel.extend({
+	defaults:_.extend(_.result(this, 'widgets.DOMWidgetModel.prototype.defaults'),{
+		_model_name: 'sourceCredChartModel',
+		_view_name: 'sourceCredChartView',
+		_model_module: 'pySourceCredChart',
+		_view_module: 'pySourceCredChart',
+	})
+});
+
+var sourceCredChartView = widgets.DOMWidgetView.extend({
+	render: function(){
+		/// set up chart space without data
+		this.chartContainer = d3.select(this.el);
+		
+		//preserve context for closures
+		var that = this;
+		
+		// define dimensions
+		this.width = 800;
+		this.height = 600;
+		
+		// set up parent element and SVG
+		this.chartContainer.innerHTML = '';
+		this.svg = this.chartContainer.append('svg');
+		this.svg.attr('width', this.width);
+		this.svg.attr('height', this.height);
+		this.chart = this.svg
+			.append("g")
+			.attr("transform", "translate(" + this.width / 2 + "," + this.height / 2 + ")")
+			
+		// set up svg defs
+		this.chart.append("defs").append("marker")
+				.attr("id", "arrow")
+				.attr("viewBox", "0 -3 10 10")
+				.attr("refX", 18)
+				.attr("refY", 0)
+				.attr("markerWidth", 5)
+				.attr("markerHeight", 5)
+				.attr("orient", "auto")
+				.append("svg:path")
+				.attr("d", "M 0,-5 L 10 ,0 L 0,5");   
+		
+		//chart elements        
+		this.edge = this.chart.append("g");
+		this.node = this.chart.append("g");
+		this.tooltip = d3.select(this.element).append("div")
+			.attr("class", "toolTip")
+			.style('display', 'none')
+			.style('position', 'absolute')
+			.style('min-width' , '50px')
+			.style('height', 'auto')
+			.style('background', 'none repeat scroll 0 0 #ffffff')
+	
+	this.model.on('change:_model_msg', this.message_changed, this);
+	this.model.on('change:_model_data', this.data_changed, this);
+	},
+	
+	message_changed: function(){
+		this.el.msg = this.model.get('_model_msg');
+		this.draw(this.el.msg[0]);
+	},
+	
+	data_changed: function(){
+		this.el.data = this.model.get('_model_data');
+		this.update(this.el.data);
+	},
+	draw: function(opts){
+		console.log(opts);
+		this.data = opts.data;
+		this.scores = opts.scores;
+		this.mapping = opts.mapping;
+		this.options = opts.options;
+		console.log(this.data);
+		console.log(this.mapping);
+		
+		//preserve context for closures
+		var that = this;
+		/*
```

Comment:
let's clean out these large blocks of commented code

Reply:
Yup